module/os/linux/zfs/zio_crypt.c

```diff
@@ -229,7 +229,14 @@ zio_crypt_key_init(uint64_t crypt, zio_crypt_key_t *key)
 	ASSERT(key != NULL);
 	ASSERT3U(crypt, <, ZIO_CRYPT_FUNCTIONS);
 
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
+#endif
 	keydata_len = zio_crypt_table[crypt].ci_keylen;
+#if defined(__GNUC__) && !defined(__clang__)
```

Comment:
I am especially curious what complaint is present here. My feeling is that it would merit a bug report with GCC.

Reply:
Is this building in userspace? My guess is that it sees the assertion and then realizes that the assertion can return when tripped due to a hack we have in place to allow assertions to be turned off in zdb. In that case, this heuristic is correct about potential out of bounds array access since our check to protect against it is not a real check. We could probably avoid this by putting the assertion in an ifdef so that it only is used when _KERNEL is defined. If that works, a comment explaining why we are doing that would be merited.

However, I am confused why I am not seeing this on Gentoo. My glibc version and GCC version should be the same as what Ubuntu 22.10 uses.

Reply:
The bug applies only to userspace.

```
Package: gcc
Version: 4:12.2.0-1ubuntu1
Package: libc6
Version: 2.36-0ubuntu4
Package: clang
Version: 1:15.0-55.1ubuntu1
Package: lld
Version: 1:15.0-55.1ubuntu1
```

Reply:
I installed a Ubuntu 22.10 userland locally, but cannot reproduce these issues. My versions for gcc and libc6 are:

```
root@vserver:~/zfs# dpkg-query --show gcc libc6
gcc     4:12.2.0-1ubuntu1
libc6:amd64     2.36-0ubuntu4
```

I did not bother with Clang since the parts of this PR that piqued my interest only apply to GCC.

More details are in the comment here:

https://github.com/openzfs/zfs/pull/14150#issuecomment-1309849929